firedrake/trefftz.py

```diff
@@ -0,0 +1,231 @@
+"""
+This module provides a class to compute the Trefftz embedding of a given function space.
+It is also used to compute aggregation embedding of a given function space.
+"""
+from firedrake.petsc import PETSc
+from firedrake.cython.dmcommon import FACE_SETS_LABEL, CELL_SETS_LABEL
+from firedrake.assemble import assemble
+from firedrake.mesh import Mesh
+from firedrake.functionspace import FunctionSpace
+from firedrake.function import Function
+from firedrake.ufl_expr import TestFunction, TrialFunction
+from firedrake.constant import Constant
+from ufl import dx, dS, inner, jump, grad, dot, CellDiameter, FacetNormal
+import scipy.sparse as sp
+
+
+class TrefftzEmbedding(object):
+    """
+    This class computes the Trefftz embedding of a given function space
+    Parameters
+    ----------
+    V : :class:`.FunctionSpace`
+        Ambient function space.
+    b : :class:`.ufl.form.Form`
+        Bilinear form defining the Trefftz operator.
+    dim : int, optional
+        Dimension of the embedding.
+        Default is the dimension of the function space.
+    tol : float, optional
+        Tolerance for the singular values cutoff.
+        Default is 1e-12.
+    backend : str, optional
+        Backend to use for the computation of the SVD.
+        Default is "scipy".
+    """
+    def __init__(self, V, b, dim=None, tol=1e-12, backend="scipy"):
+        self.V = V
+        self.b = b
+        self.dim = V.dim() if not dim else dim + 1
+        self.tol = tol
+        self.backend = backend
+
+    def assemble(self):
+        """
+        Assemble the embedding, compute the SVD and return the embedding matrix
+        """
+        self.B = assemble(self.b).M.handle
+        if self.backend == "scipy":
```

Comment:
Does this work in parallel? We should have a parallel test.

Reply:
Not sure having a backend way we want to do this either...